File: 백준/Gold/11265. 끝나지 않는 파티/끝나지 않는 파티.py
from sys import stdin

input = stdin.readline

## 변수 입력 부분 ##
n, m = map(int, input().split())  # n: 파티장의 크기, m: 요청한 손님의 수
matrix = [list(map(int, input().split())) for _ in range(n)]  # (i, j) : i -> j 직접적으로 걸리는 시간

# 플로이드 워셜 점화식 구현
for k in range(n):
    for i in range(n):
        for j in range(n):
            if matrix[i][j] > matrix[i][k] + matrix[k][j]:
                matrix[i][j] = matrix[i][k] + matrix[k][j]

for _ in range(m):
    A, B, C = map(int, input().split())  # A: 현재 파티장 번호, B: 다음 파티장 번호, C: 파티가 열리는 시간
    if matrix[A - 1][B - 1] <= C:
        print("Enjoy other party")
    else:
        print("Stay here")

File: Java/백준/Gold/11265. 끝나지 않는 파티/끝나지 않는 파티.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {

  private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
  private static final StringBuilder sb = new StringBuilder();
  private static StringTokenizer st;

  private static int n, m, a, b, c;
  private static int[][] map;
  private static int[][] dist;

  public static void main(String[] args) throws Exception {
    init();
    solve();
  }

  private static void init() throws Exception {

    st = new StringTokenizer(br.readLine());
    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());

    map = new int[n][n];
    for (int i = 0; i < n; i++) {
      st = new StringTokenizer(br.readLine());
      for (int j = 0; j < n; j++) {
        map[i][j] = Integer.parseInt(st.nextToken());
      }
    }
  }

  private static void solve() throws IOException {

    floydWarshall();

    for (int i = 0; i < m; i++) {
      st = new StringTokenizer(br.readLine());
      a = Integer.parseInt(st.nextToken());
      b = Integer.parseInt(st.nextToken());
      c = Integer.parseInt(st.nextToken());
      if (dist[a - 1][b - 1] <= c) {
        sb.append("Enjoy other party").append("\n");
      } else {
        sb.append("Stay here").append("\n");
      }
    }

    System.out.print(sb);
    br.close();
    sb.setLength(0);
  }

  private static void floydWarshall() {

    final int INF = 1_000_000_001;
    dist = new int[n][n];
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        if (i == j) {
          dist[i][j] = 0;
        } else if (map[i][j] == 0) {
          dist[i][j] = INF;
        } else {
          dist[i][j] = map[i][j];
        }
      }
    }

    for (int k = 0; k < n; k++) {
      for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
          if (dist[i][j] > dist[i][k] + dist[k][j]) {
            dist[i][j] = dist[i][k] + dist[k][j];
          }
        }
      }
    }
  }
}